### PR TITLE
add support for push_subscriptions (partner API)

### PR DESCRIPTION
**Pull request by YousefED**
docs @ strava.github.io/api/partner/v3/events/

### CODE REVIEW

**Comment by UnbounDev** [comment]
Thanks for the work, @YousefED - before merging and publishing can you please add an analogous `test/push_subscriptions.js` mocha test for the `push_subscriptions` functions?

**Comment by YousefED** [comment]
Can do, but they would all return "unauthenticated" since this is a private API

**Comment by TimoSta** [comment]
@YousefED This looks good to me.

I can absolutely see your point regarding the test cases. Writing unit tests that are bound to fail due to access restrictions are pointless and would have to be skipped anyway, since our Travis build would fail otherwise. Considering that we also have a problem with a pretty severe rate limit (max. 600 requests per 15 mins) for our free test account, running our unit test suite against the actual API is not really viable going forward.

For this reason, I'm planning to write a stub for the API that we can use for our unit tests, and only run select integration tests against the actual Strava API. For now, I will merge this feature into the codebase without a dedicated unit test. Once we are able to properly execute test cases, we should add a test case for this endpoint, as well as for several others that are not completely covered or not covered at all at the moment.

One small nit-pick before we go ahead with the merge, could you please change the identifier `push_subscriptions` to camel case `pushSubscriptions`? This would be in line with common JavaScript conventions, and the rest of the code base.

Thank you!

**Comment by YousefED** [comment]
@TimoSta not tested, but this should do it: https://github.com/YousefED/node-strava-v3/blob/YousefED-patch-1/

**Comment by TimoSta** [comment]
@YousefED Not sure what you are referring to, the link you posted seems to be dead?

Could you please add the change to the pull request here?
Thank you!

**Comment by markstos** [comment]
@YousefED I'm interested in this functionality as well. Let me know if you need any help updating the pull request as requested.

**Comment by YousefED** [comment]
@markstos would be helpful! I did some quickfixes just now, but not passing yet. If you have time, it would be great if you could:

- make sure the build passes
- Actually test pushSubscriptions in your own project. As our push subscription has been running for a while, I haven't tested this code (and I don't want to fiddle with our subscription at this moment). Hope this makes sense :)

**Comment by markstos** [comment]
@TimoSta Regarding testing against mock data vs the real API-- that is what Strava does in their own REST client, written in Go.  They have even nicely separated out their mock responses and body's as JSON objects, which could easily be copied into this project. Find that data here: https://github.com/strava/go.strava/tree/master/cassettes

The Subscriptions feature could be a great feature to try out the approach on.

**Comment by markstos** [comment]
I've realized I don't actually need this initially. We are going to create our webhook subscriptions manually for now and don't need the other methods.

I've reviewed the Travis failures on this branch, and they are failing due to a missing access token of some kind. It's not clear that this issue is caused by this branch and not some other issue. I would encourage other reviewers to pull down this branch locally and see if it passes for them locally.

File: lib/pushSubscriptions.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+pushSubscriptions.endpointBase = 'https://api.strava.com/api/v3/';
+
+pushSubscriptions.post = function(args,done) {
+    var endpoint = 'push_subscriptions';
+    var err = null;
+
+    if(
+       typeof args.object_type === 'undefined' || 
+       typeof args.aspect_type == 'undefined' || 
+       typeof args.callback_url == 'undefined'
```
Comment (markstos) [inline lib/pushSubscriptions.js]: Why does the first check use "triple equals" to check against `undefined`, but the second two checks only use a "double equals" check against `undefined` ?

**Comment by markstos** [comment]
It turns out that I *do* want to use the  Webhook events feature in this API, so I'm looking at merging this again today.

**Comment by markstos** [comment]
I now have a version of Push Subscription support that adds test coverage and has all regression tests passes. It also documents the feature in the README. It works with both the new Promise-based API I added as well as the callback API. The commit is here: https://github.com/RideAmigosCorp/node-strava-v3/commit/be812221055cbfd6c9127ef476dc9aa264fd9331

The caveat is that is the code is based on the increasingly large pile of unmerged contributions I've made.

If you want to checkout a fully working tree that includes the feature, see the `ra-fixes` branch in the RideAmigosCorp repo:

https://github.com/RideAmigosCorp/node-strava-v3/tree/ra-fixes

In this branch I also introduce the use of `nock` to mock the Strava API. These tests run *much* faster. The mocked tests also don't require a multi-step setup to get a working `access_token` nor do they require particular pre-existing data in a related Strava account.  I think mocking the Strava API in more places would make development easier for more people to contribute, too. Right now, it's a bit tedious to setup a development environment with all tests passing.

**Comment by markstos** [comment]
I made a couple more changes / improvements in my markstos/support-push-subscriptions branch:

 - I renamed post() to create() on `strava.pushSubscriptions`. create() is more consistent with activity.create()
- I set object_type and aspect_type default values when creating a push subscription. There's no need to require people to pass in the object_type and aspect_type when there is only one valid value.

**Comment by vlavrynovych** [comment]
@markstos @TimoSta Hi! Any updates on this?

**Comment by markstos** [comment]
No. My numerous proposed contributions including this feature have been ignored. At this point it's probably best to fork the project and release the updates under a new name, or ask the project owners here if they would take on co-maintainers. I could help a little, but don't have time to be a lead maintainer of another project now.

**Comment by markstos** [comment]
I'm closing this because I'm now a maintainer and plan to merge a version of my "ra-fixes" branch which includes webhook event support.

**Pull request closed without merging**